WooCommerce/Classes/ViewRelated/Orders/Order Details/Issue Refunds/IssueRefundViewController.swift

```diff
@@ -17,6 +17,14 @@ final class IssueRefundViewController: UIViewController {
 
     private let viewModel: IssueRefundViewModel
 
+    /// Closure invoked when the next button is tapped
+    ///
+    var onNextAction: ((RefundConfirmationViewModel) -> Void)?
```

Comment:
I am aware we are already using this pattern is another few view controllers, but I am not sure having view controllers expose a closure directly is the best way to decouple UI from navigation. 

I would argue that it couples things (view controller and navigation controller) even more. In this case in particular, the navigation controller  (coordinator) would need to be aware of the existence of RefundConfirmationViewController. Which might be necessary to instantiate the next step in the navigation sequence, but still, I would argue should not be knowledge of a navigation controller.

Would it make sense to encapsulate these closures in an object, and pass an instance of that object instead? Sort of a lightweight implementation of the command pattern, where what we pass to the view controller is the implementation of an interface with just one method (say, `execute()`). 

Or probably better, we could have a different "NavigationCommands" object per view controller, that would pack all navigation out of a given view controller, and notify clients via a delegate.

That way we would completely decouple view controllers from navigation controllers. 

On the other hand, that would mean that we would be adding a significant amount of very granular objects.

But at the same time, each of everyone of those objects would be testable as well, so we would be able to test:
* that the view controller meets expectations when navigation buttons are tapped
* that the entities that communicate with the navigation controller meet expectations.

On the other hand, I would argue that a simplest solution, that does not involve navigation controllers (i.e. coordinators) would be making viewmodels responsible for triggering navigation. In a way, a view controller should be already calling into a view model when there s user interaction in the UI...

What do you think?

Reply:
> I would argue that it couples things (view controller, view model and coordinator) even more. In this case in particular, the coordinator would need to be aware of the existence of RefundConfirmationViewController. Which might be necessary to instantiate the next step in the navigation sequence, but still, I would argue should not be knowledge of a navigation controller.

Maybe we understand the coordinator pattern differently but I would say that the `Coordinator` does need to know about what view controller is coming next, since the coordinator's job is to define what VC comes next in the flow.

> Would it make sense to encapsulate these closures in an object, and pass an instance of that object instead? Sort of a lightweight implementation of the command pattern, where what we pass to the view controller is the implementation of an interface with just one method (say, execute()).

I'm not too sure, that object would have to be created by the `ViewController` or the `ViewModel` right? If that is the case then the `ViewController` couldn't really be reused in a different flow without adding conditionals all over the place. I think that defeats the main purpose of the `Coordinator`.

Reply:
FYI you can see here(p91TBi-36A-p2) the discussion and rationale when this pattern was adopted.

Reply:
The way I see it, the coordinator pattern is just the [Application Controller](https://martinfowler.com/eaaCatalog/applicationController.html) pattern, as defined in [Patterns of Enterprise Application Architecture](https://www.amazon.com/Patterns-Enterprise-Application-Architecture-Martin/dp/0321127420)

The spirit of the pattern is `A centralized point for handling screen navigation and the flow of an application.`, which is what this PR implements. That's a 💯

Yes, the Application Controller is responsible for instantiating and navigating to the next step in the sequence, encapsulating the logic to execute those actions.

What I don't understand is:
> If that is the case then the ViewController couldn't really be reused in a different flow without adding conditionals all over the place. I think that defeats the main purpose of the Coordinator.

Why would we want to reuse the view controller in a different flow?  I am genuinely curious. Assuming we wanted to reuse it in a different flow, decoupling the view controller from the application controller as much as possible seems like the safest way to do it.

Reply:
Some use cases I can think of would be:
- A/B Testing: Having different coordinators served to a different set of people.
- iPad Coordinators: IPad apps may choose to display or handle view controllers in a different way. Having VC's reusable will help us to achieve that.